src/renderer/services/thorchain/types.ts

```diff
@@ -70,3 +75,81 @@ export type Mimir = t.TypeOf<typeof MimirIO>
 
 export type MimirLD = LiveData<Error, Mimir>
 export type MimirRD = RD.RemoteData<Error, Mimir>
+
+export type GetLiquidityProvidersParams = {
+  asset: Asset
+  network: Network
+  assetDecimal: number
+}
+
+export type GetLiquidityProviderParams = GetLiquidityProvidersParams & {
+  walletAddress: Address
+}
+
+export type LiquidityProvider = {
+  asset: Asset
+  address: Address
+  pendingRune: BaseAmount
+  pendingAsset: BaseAmount
+  runeDepositValue: BaseAmount
+  assetDepositValue: BaseAmount
+}
+
+export type LiquidityProvidersLD = LiveData<ApiError, LiquidityProvider[]>
+export type LiquidityProviderLD = LiveData<ApiError, O.Option<LiquidityProvider>>
+export type LiquidityProviderRD = RD.RemoteData<ApiError, O.Option<LiquidityProvider>>
+
+const assetDecoder: D.Decoder<unknown, Asset> = FP.pipe(
+  D.string,
+  D.parse((stringAsset) =>
+    FP.pipe(
+      stringAsset,
+      assetFromString,
+      E.fromNullable(new Error('Invalid asset')),
+      E.fold((e) => D.failure(stringAsset, e.message), D.success)
+    )
+  )
+)
+
+/**
+ * Custom type validator to check if string is Asset instanse
+ * After successful validation transforms string to Asset
+ */
+const assetType = new t.Type(
+  'asset type',
+  (input): input is Asset => {
+    if (typeof input === 'string') {
+      const asset = assetFromString(input)
+
+      if (asset) {
+        return isValidAsset(asset)
+      }
+    }
+    return false
+  },
```

Comment:
Why do we need two checks for `input === string` (here ^ and in next step)?

Reply:
cuz initially `input` has `unknown` type and we strictly check for this value to be a string. the main idea of all codecs of `io-ts` is that you always getting `unknown` value and do all the checks you need by hands

Reply:
Make sense.